qualtran/qref_interop/_bloq_to_qref.py

```diff
@@ -0,0 +1,311 @@
+import networkx as nx
+
+from functools import singledispatch
+from typing import Any, Iterable, Optional, Union
+
+import sympy
+from qref.schema_v1 import PortV1, RoutineV1, SchemaV1
+
+from qualtran import Bloq, BloqInstance, CompositeBloq
+from qualtran import Connection as QualtranConnection
+from qualtran import Register, Side, Soquet
+from qualtran.cirq_interop import CirqGateAsBloq
+
+
+@singledispatch
+def _bloq_type(bloq: Bloq) -> str:
+    """Determine type of Bloq.
+
+    Output of this function is used for populating the `type` field of RoutineV1
+    build from imported Bloqs. By default, we use name of the type of the Bloq.
+    """
+    return type(bloq).__name__
+
+
+@_bloq_type.register
+def _cirq_gate_bloq_type(bloq: CirqGateAsBloq) -> str:
+    """Determine type of Bloq constructed from Cirq gate.
+
+    Without this variant of _bloq_type, type of all instances of CirqGateAsBloq
+    are set to "CirqGateAsBloq" which is not helpful, because all gates would get the
+    same type. Instead of this default behaviour, we extract textual representation
+    of the Cirq's gate.
+    """
+    return str(bloq.gate)
+
+
+def _is_symbol_or_int(expression):
+    try:
+        int(expression)
+        return True
+    except (TypeError, ValueError):
+        return expression.isidentifier()
+
+
+def _extract_common_bloq_attributes(bloq: Bloq, name: Optional[str] = None) -> dict[str, Any]:
+    """Extract common bloq attributes such as name, type and ports.
+
+    There are several Bloq classes, however, they all share common set of atributes.
+    This function is used to extract them, so that we don't have to duplicate logic
+    for each Bloq subtype.
+
+    Args:
+        bloq: Bloq for which information should be extracted.
+        name: Optional override for name. Setting it to None will cause some reasonable
+            name (depending on bloq type) to be inferred. The typical use of non-trivial
+            name is when a better name is known from the parent Bloq.
+
+    Returns:
+        A dictionary that can be unpacked into arguments of RoutineV1 initializer.
+    """
+    ports = [port for reg in bloq.signature for port in _ports_from_register(reg)]
+    local_variables = {}
```

Comment:
Can you add a comment explaining what's the use case for `local_variables` ? It seems like it may be used when the string representation of the symbolic size of a qualtran register does not satisfy the `isIdentifier()` condition which is needed for any string to be a valid port size in QREF ?

Reply:
After some thoughts I decided this logic should actually go to `bartiq`, will update `bartiq` and then remove it from here.

Reply:
Actually, it would require some major changes in bartiq and if done like this it should work, so I'll leave it as it is.
Added a comment.

Reply:
An extra question – is there a way to get all the symbols used in a particular `bloq`? 
In most cases the symbols taken from `signature` are enough, but sometimes (e.g. in `PermutationCycle`), I don't know how to retrieve the information that it requires having `L` defined, other than parsing the `t_complexity` of it.

Reply:
We don't have an interface that does exactly this. We often try to treat symbols and concrete values as similarly as possible in the code (this was our design decision; it has its plus-es and minus-es). In practice, this means symbols get passed in as constructor arguments (aka bloq attributes). But they can be arbitrarily hidden: 

```python
    N, L = sympy.symbols("N L", positive=True, integer=True)
    cycle = Shaped((L,))
    permutation_cycle_symb = PermutationCycle(N, cycle)
```

the `L` symbol is passed as a constructor argument, but contained within a `Shaped` object (which we use to support cases where the concrete value is an array but we only need the shape to do e.g. resource counts).

Since everything is an `attrs` dataclass, you can extract the bloq attributes with (pseudocode)

```python
fields = attrs.fields(bloq)
vals = {f.name: getattrs(bloq, f.name) for f in fields}
```

so you can see that the vals dictionary contains `{'N': sympy.Symbol('N')` (easy), and `'cycle': Shaped(sympy.Symbol('L'))}` (you have to poke at it to see that it has a symbol named L